staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -235,144 +213,136 @@ const (
-// irrelevant.
-func (qs *queueSet) Wait(ctx context.Context, hashValue uint64, descr1, descr2 interface{}) (tryAnother, execute bool, afterExecution func()) {
+// StartRequest begins the process of handling a request.  We take the
+// approach of updating the metrics about total requests queued and
+// executing on each path out of this method and Request::Wait.  We do
+// not update those metrics in lower level functions because there can
+// be multiple lower level changes in one invocation here.
+func (qs *queueSet) StartRequest(ctx context.Context, hashValue uint64, descr1, descr2 interface{}) (fq.Request, bool) {
+	qs.lockAndSyncTime()
+	defer qs.lock.Unlock()
 	var req *request
-	decision := func() requestDecision {
-		// Decide what to do and update metrics accordingly.  Metrics
-		// about total requests queued and executing are updated on
-		// each way out of this function rather than in lower level
-		// functions because there can be multiple lower level changes
-		// in one invocation here.
-		qs.lockAndSyncTime()
-		defer qs.lock.Unlock()
-		// A call to Wait while the system is quiescing will be rebuffed by
-		// returning `tryAnother=true`.
-		if qs.emptyHandler != nil {
-			klog.V(5).Infof("QS(%s): rebuffing request %#+v %#+v with TryAnother", qs.qCfg.Name, descr1, descr2)
-			return decisionTryAnother
+
+	// ========================================================================
+	// Step 0:
+	// Apply only concurrency limit, if zero queues desired
+	if qs.qCfg.DesiredNumQueues < 1 {
+		if qs.totRequestsExecuting >= qs.dCfg.ConcurrencyLimit {
+			klog.V(5).Infof("QS(%s): rejecting request %#+v %#+v because %d are executing and the limit is %d", qs.qCfg.Name, descr1, descr2, qs.totRequestsExecuting, qs.dCfg.ConcurrencyLimit)
+			return nil, qs.isIdleLocked()
 		}
+		req = qs.dispatchSansQueueLocked(ctx, descr1, descr2)
+		metrics.UpdateFlowControlRequestsExecuting(qs.qCfg.Name, qs.totRequestsExecuting)
+		return req, false
+	}
+
+	// ========================================================================
+	// Step 1:
+	// 1) Start with shuffle sharding, to pick a queue.
+	// 2) Reject old requests that have been waiting too long
+	// 3) Reject current request if there is not enough concurrency shares and
+	// we are at max queue length
+	// 4) If not rejected, create a request and enqueue
+	req = qs.timeoutOldRequestsAndRejectOrEnqueueLocked(ctx, hashValue, descr1, descr2)
+	defer metrics.UpdateFlowControlRequestsInQueue(qs.qCfg.Name, qs.totRequestsWaiting)
+	// req == nil means that the request was rejected - no remaining
+	// concurrency shares and at max queue length already
+	if req == nil {
+		klog.V(5).Infof("QS(%s): rejecting request %#+v %#+v due to queue full", qs.qCfg.Name, descr1, descr2)
+		metrics.AddReject(qs.qCfg.Name, "queue-full")
+		return nil, qs.isIdleLocked()
+	}
+
+	// ========================================================================
+	// Step 2:
```

Comment:
I'm unclear on why this step is required? We're running under the lock, so between the start of this function and now, the only change that could have happened to the concurrency level is that, if it was below the threshhold, it could have gone up by one in step 1 above (in which case there are perforce no non-empty queues). If there was anything queued, then our concurrency limit is full, and the executing requests' afterExecute functions should dequeue the next requests.

What am I missing?

Reply:
Step 1 only puts the request in a queue; step 2 can take the request out of that queue and start execution.

What we know at this point, due to the lock covering both steps 1 and 2, is that there is at least one request in the queue.  If we add knowledge about how reconfigurations are handled, we also know that at most one request can be dispatched at this point.  We _could_ add a special case dispatch-at-most-1 function, but I think it is simpler to just re-use the dispatch-as-many-as-possible function.